File: runtime/ruby/units/wasi/fd_filestat_set_times.rb
# requires: wasi/set_times, wasi/errno_fs
def wasi_fd_filestat_set_times(fd, atim, mtim, fstflags)
  io = @fds[fd]
  return ERRNO_BADF if io.nil? || io.is_a?(WasiDir)
  err = validate_fstflags(fstflags)
  return err if err
  a, m = resolve_times(io.stat, atim, mtim, fstflags)
  File.utime(a, m, io.path)
  ERRNO_SUCCESS
rescue SystemCallError => e
  fs_errno(e)
end

File: runtime/perl/units/wasi/fd_filestat_set_times.pl
# requires: wasi/fst_times
use Time::HiRes ();

sub wasi_fd_filestat_set_times {
    my ($self, $fd, $atim, $mtim, $fst_flags) = @_;
    my $e = $self->{fds}{$fd};
    return ERRNO_BADF if !defined($e) || $e->{dir};
    my @st = Time::HiRes::stat($e->{fh});
    return ERRNO_IO unless @st;
    my ($a, $m, $err) = $self->fst_times($st[8], $st[9], $atim, $mtim, $fst_flags);
    return $err if defined $err;
    Time::HiRes::utime($a, $m, $e->{fh}) or return ERRNO_IO;
    return ERRNO_SUCCESS;
}
